### PR TITLE
Add minimumBy/maximumBy/Option to Foldable

### LINKED ISSUE #2385: Add Foldable#[max|min]ByOption, Reducible#[maxBy|minBy]

### PR DESCRIPTION
**Pull request by joroKr21**
* `mimimumByOption` / `maximumByOption` on `Foldable`
  * `minumumBy` / `maximumBy` on `Reducible`

Thank you for contributing to Cats!

Fixes #2385

### CODE REVIEW

**Comment by kailuowang** [comment]
Let's add these to the type class. Now that we dropped 2.11 on master it's BC to add methods to trait. The syntax addition can be a PR to the Scala_2.11 branch

**Comment by joroKr21** [comment]
Looks like unrelated failure:
```
[info] - CommutativeGroup[BigDecimal].commutativeGroup.associative *** FAILED ***
[info]   GeneratorDrivenPropertyCheckFailedException was thrown during property evaluation.
[info]    (Discipline.scala:12)
[info]     Falsified after 39 successful property evaluations.
[info]     Location: (Discipline.scala:12)
[info]     Occurred when passed generated values (
[info]       arg0 = -4.420570652140785E+214,
[info]       arg1 = -5.873313741220067E+186,
[info]       arg2 = 5.83376395628893E+188
[info]     )
[info]     Label of failing property:
[info]       Expected: -4.420570652140784999999999942249692E+214
[info]   Received: -4.420570652140784999999999942249691E+214
```

File: tests/src/test/scala/cats/tests/FoldableSuite.scala [hunk 1]
```diff
@@ -192,6 +192,20 @@ abstract class FoldableSuite[F[_]: Foldable](name: String)(implicit ArbFInt: Arb
     }
   }
 
+  test(s"Foldable[$name].maximumBy/minimumBy") {
```
Comment (kailuowang) [inline tests/src/test/scala/cats/tests/FoldableSuite.scala]: sorry I should've mentioned this in the earlier comment. Can we make these laws like the other `_Ref` laws in Foldable Laws and Reducible laws?

Comment (joroKr21) [reply tests/src/test/scala/cats/tests/FoldableSuite.scala]: Hmm I'm not sure what you mean :thinking: couldn't find `_Ref` anywhere in the file.

Comment (joroKr21) [reply tests/src/test/scala/cats/tests/FoldableSuite.scala]: Ah ok I found them. Ok I will add some laws. Should I keep the tests in the suite?

Comment (kailuowang) [reply tests/src/test/scala/cats/tests/FoldableSuite.scala]: I think having them in law is enough. sorry I should've been clear about where the law code is.

Comment (joroKr21) [reply tests/src/test/scala/cats/tests/FoldableSuite.scala]: Hmm, if I want to add tests about minimum/maximum to `FoldableTests` that would mean adding an `Order` constraint on `B` in `def foldable[A: Arbitrary, B: Arbitrary]`

Comment (kailuowang) [reply tests/src/test/scala/cats/tests/FoldableSuite.scala]: Ah, right. Then we can't add to laws. We can only use the current tests
